client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -554,75 +593,133 @@ private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
+     * that. If the retries fail this throws a {@link IOException}.
+     * @throws IOException if no nodes are available
      */
-    private HostTuple<Iterator<HttpHost>> nextHost() {
-        final HostTuple<Set<HttpHost>> hostTuple = this.hostTuple;
-        Collection<HttpHost> nextHosts = Collections.emptySet();
-        do {
-            Set<HttpHost> filteredHosts = new HashSet<>(hostTuple.hosts);
-            for (Map.Entry<HttpHost, DeadHostState> entry : blacklist.entrySet()) {
-                if (entry.getValue().shallBeRetried() == false) {
-                    filteredHosts.remove(entry.getKey());
-                }
+    private NodeTuple<Iterator<Node>> nextNode(NodeSelector nodeSelector) throws IOException {
+        NodeTuple<List<Node>> nodeTuple = this.nodeTuple;
+        List<Node> hosts = selectHosts(nodeTuple, blacklist, lastNodeIndex, System.nanoTime(), nodeSelector);
+        return new NodeTuple<>(hosts.iterator(), nodeTuple.authCache);
+    }
+
+    /**
+     * Select hosts to try. Package private for testing.
+     */
+    static List<Node> selectHosts(NodeTuple<List<Node>> nodeTuple,
+            Map<HttpHost, DeadHostState> blacklist, AtomicInteger lastNodeIndex,
+            long now, NodeSelector nodeSelector) throws IOException {
+        /*
+         * Sort the nodes into living and dead lists.
+         */
+        List<Node> livingNodes = new ArrayList<>(nodeTuple.nodes.size() - blacklist.size());
+        List<DeadNodeAndRevival> deadNodes = new ArrayList<>(blacklist.size());
+        for (Node node : nodeTuple.nodes) {
+            DeadHostState deadness = blacklist.get(node.getHost());
+            if (deadness == null) {
+                livingNodes.add(node);
+                continue;
             }
-            if (filteredHosts.isEmpty()) {
-                //last resort: if there are no good hosts to use, return a single dead one, the one that's closest to being retried
-                List<Map.Entry<HttpHost, DeadHostState>> sortedHosts = new ArrayList<>(blacklist.entrySet());
-                if (sortedHosts.size() > 0) {
-                    Collections.sort(sortedHosts, new Comparator<Map.Entry<HttpHost, DeadHostState>>() {
-                        @Override
-                        public int compare(Map.Entry<HttpHost, DeadHostState> o1, Map.Entry<HttpHost, DeadHostState> o2) {
-                            return o1.getValue().compareTo(o2.getValue());
-                        }
-                    });
-                    HttpHost deadHost = sortedHosts.get(0).getKey();
-                    logger.trace("resurrecting host [" + deadHost + "]");
-                    nextHosts = Collections.singleton(deadHost);
+            long nanosUntilRevival = now - deadness.getDeadUntilNanos();
+            if (nanosUntilRevival > 0) {
+                livingNodes.add(node);
+                continue;
+            }
+            deadNodes.add(new DeadNodeAndRevival(node, nanosUntilRevival));
```

Comment:
maybe call the class DeadNodeState ?

Reply:
I don't want to clash with `DeadHostState`. They sort of do different things too.